src/Symfony/Bundle/FrameworkBundle/Secret/SodiumEncoder.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Secret;
+
+use Symfony\Bundle\FrameworkBundle\Exception\EncryptionKeyNotFoundException;
+use Symfony\Component\Filesystem\Filesystem;
+
+/**
+ * @author Tobias Schultze <http://tobion.de>
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class SodiumEncoder implements EncoderInterface
+{
+    private $encryptionKey;
+    private $encryptionKeyOrPath;
+
+    public function __construct(string $encryptionKeyOrPath)
+    {
+        if (!\function_exists('\sodium_crypto_stream_xor')) {
+            throw new \RuntimeException('The "sodium" PHP extension is not loaded.');
+        }
+
+        $this->encryptionKeyOrPath = $encryptionKeyOrPath;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generateKeys(bool $override = false): array
+    {
+        $path = $this->getKeyPath();
+        if (!$override && \is_file($path)) {
+            throw new \LogicException(sprintf('A key already exist in "%s".', $path));
+        }
+
+        $this->encryptionKey = null;
+
+        $encryptionKey = \sodium_crypto_stream_keygen();
+        (new Filesystem())->dumpFile($path, $encryptionKey);
+        \sodium_memzero($encryptionKey);
+
+        return [$path];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function encrypt(string $message): string
+    {
+        $nonce = random_bytes(\SODIUM_CRYPTO_STREAM_NONCEBYTES);
+
+        $key = $this->getKey();
+        $encryptedText = \sodium_crypto_stream_xor($message, $nonce, $key);
+        \sodium_memzero($message);
+        \sodium_memzero($key);
+
+        return $this->encode($nonce, $encryptedText);
+    }
+
+    private function getKeyPath(): ?string
+    {
+        if (null === @\is_file($this->encryptionKeyOrPath)) {
+            throw new \InvalidArgumentException('The path to the encryption key is not valid. Did you configured the framework with the content of key instead of its location?');
+        }
+
+        return $this->encryptionKeyOrPath;
+    }
+
+    private function getKey(): string
+    {
+        if (isset($this->encryptionKey)) {
+            return $this->encryptionKey;
+        }
+        // is_file returns null when the argument is not a valid string, meaning `encryptionKeyOrPath` contains the key itself
```

Comment:
you cannot rely on this. it depends on what chars the encryption key contains which is random. if it contains non-filesystem chars like null-byte, then it returns null. but if non, it can also return false:
https://3v4l.org/Q33Xq

Reply:
that was the purpose of fallback to is_file==true below, , then check the length of the key.

TBH: I'm not very confortable with this part of code too (BTW, LexikJWTAuthenticationBundle
 uses the same thinkg). This allow user to provide either a path to the KEY or the content of the key. 

I think we should only allow path to key, It would simplify the code a lot

Reply:
I would indeed be explicit and only accept a file path.

Reply:
I would also prefer if we only need to support a file path. But I don't see how to set the encryption key for most deployments then. Normally I would set the encryption key as environment variable, e.g. in cloudfoundry or gitlab ci, which is very easy. And then configure the encryption as to read it from the env variable. Very straightforward and no extra deployment steps needed.
I don't see how to do that with a file. I cannot create a file as an env variable. So this would require an extra deployment script that generates the file or something.

Reply:
Note for gitlab: env variable are easily dumpable in the build script. (nb: this is what gitlab's documentation describes when using SSH https://docs.gitlab.com/ee/ci/ssh_keys/#ssh-keys-when-using-the-docker-executor)

Managing files is also the way to go with docker swarm and kubernetes secrets.

IMHO: Allowing env variables could be done in an other PR.
By the meantime, users could decorate the SodiumEncoder or implement a `stream` EnvResolver that act like the opposite of `file`: convert a string into a resource